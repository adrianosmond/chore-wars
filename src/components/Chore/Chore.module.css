.divider {
  display: flex;
}

.left {
  flex-grow: 1;
  padding-right: 2rem;
}

.progressBar {
  display: flex;
  align-items: flex-end;
  height: 2.5rem;
  margin-bottom: var(--spacing-s);
}

.right {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: flex-end;
}

.date {
  margin-top: var(--spacing-xs);
  font-size: 0.75rem;
  text-align: right;
}

.overdue {
  color: var(--color-warning);
}

.completeButton {
  width: 2.5rem;
  min-width: 2.5rem;
  padding: 0.625rem;
}
