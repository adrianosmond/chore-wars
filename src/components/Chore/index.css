.chore {
  position: relative;
  padding: 1rem;
  padding-right: 5rem;
  background-color: dodgerblue;
  color: #fff;
}

.chore__title {
  font-weight: bold;
}

.chore__bar {
  position: relative;
  height: 0.25rem;
  margin-top: 2rem;
}

.chore__bar::before {
  content: "";
  position: absolute;
  top: 0;
  left: 0;
  height: 0.25rem;
  width: 100%;
  background-color: #fff;
}

.chore__bar-inner {
  position: absolute;
  bottom: 0;
  left: 0;
  height: 0.25rem;
  background-color: navy;
  text-align: center;
}

.chore__bar-inner--bonus {
  background-color: crimson;
}

.chore__points {
  position: absolute;
  z-index: 1;
  bottom: 0.5rem;
  right: 1px;
  padding: 0.25rem 0.375rem;
  font-size: 0.75rem;
  background-color: navy;
  border-radius: 0.25rem;
  transform: translateX(50%);
}


.chore__points::after {
  content: "";
  position: absolute;
  bottom: -3px;
  left: 50%;
  margin-left: -3px;
  width: 0;
  height: 0;
  border-style: solid;
  border-width: 3px 3px 0 3px;
  border-color: navy transparent transparent transparent;
}


.chore__points--bonus {
  background-color: crimson;
}

.chore__points--bonus::after {
  border-color: crimson transparent transparent transparent;
}

.chore__last-done {
  margin-top: 0.5rem;
  font-size: 0.75rem;
  text-align: right;
}

.chore__last-done--overdue {
  color: crimson;
  font-weight: bold;
}

.chore__complete-button {
  position: absolute;
  display: block;
  top: 1rem;
  right: 1rem;
  width: 2.5rem;
  height: 2.5rem;
  background-color: #fff;
  border: none;
  border-bottom: 2px solid #ccc;
  cursor: pointer;
}

.chore__complete-button:active {
  border-bottom: 0;
  border-top: 2px solid transparent;
}

.chore__complete-button:disabled {
  border-bottom: 2px solid #ccc;
  border-top: 0;
  opacity: 0.5;
  pointer-events: none;
}

.chore__extra-options {
  position: absolute;
  right: 1rem;
  top: 4rem;
}
