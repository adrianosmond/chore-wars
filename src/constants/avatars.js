export const avatarOptions = {
  skinColor: [
    'Pale',
    'Light',
    'Tanned',
    'Yellow',
    'Brown',
    'DarkBrown',
    'Black',
  ],
  topType: [
    'NoHair',
    'Eyepatch',
    'Hat',
    'Hijab',
    'Turban',
    'WinterHat1',
    'WinterHat2',
    'WinterHat3',
    'WinterHat4',
    'LongHairBigHair',
    'LongHairBob',
    'LongHairBun',
    'LongHairCurly',
    'LongHairCurvy',
    'LongHairDreads',
    'LongHairFrida',
    'LongHairFro',
    'LongHairFroBand',
    'LongHairNotTooLong',
    'LongHairShavedSides',
    'LongHairMiaWallace',
    'LongHairStraight',
    'LongHairStraight2',
    'LongHairStraightStrand',
    'ShortHairDreads01',
    'ShortHairDreads02',
    'ShortHairFrizzle',
    'ShortHairShaggyMullet',
    'ShortHairShortCurly',
    'ShortHairShortFlat',
    'ShortHairShortRound',
    'ShortHairShortWaved',
    'ShortHairSides',
    'ShortHairTheCaesar',
    'ShortHairTheCaesarSidePart',
  ],
  hatColor: [
    'Black',
    'Blue01',
    'Blue02',
    'Blue03',
    'Gray01',
    'Gray02',
    'Heather',
    'PastelBlue',
    'PastelGreen',
    'PastelOrange',
    'PastelRed',
    'PastelYellow',
    'Pink',
    'Red',
    'White',
  ],
  hairColor: [
    'Auburn',
    'Black',
    'Blonde',
    'BlondeGolden',
    'Brown',
    'BrownDark',
    'PastelPink',
    'Platinum',
    'Red',
    'SilverGray',
  ],
  facialHairType: [
    'Blank',
    'BeardMedium',
    'BeardLight',
    'BeardMagestic',
    'MoustacheFancy',
    'MoustacheMagnum',
  ],
  facialHairColor: [
    'Auburn',
    'Black',
    'Blonde',
    'BlondeGolden',
    'Brown',
    'BrownDark',
    'Platinum',
    'Red',
  ],
  accessoriesType: [
    'Blank',
    'Kurt',
    'Prescription01',
    'Prescription02',
    'Round',
    'Sunglasses',
    'Wayfarers',
  ],
  clotheType: [
    'BlazerShirt',
    'BlazerSweater',
    'CollarSweater',
    'Hoodie',
    'Overall',
    'ShirtCrewNeck',
    'ShirtScoopNeck',
    'ShirtVNeck',
  ],
  clotheColor: [
    'Black',
    'Blue01',
    'Blue02',
    'Blue03',
    'Gray01',
    'Gray02',
    'Heather',
    'PastelBlue',
    'PastelGreen',
    'PastelOrange',
    'PastelRed',
    'PastelYellow',
    'Pink',
    'Red',
    'White',
  ],
};


export const editorOrder = [
  'skinColor',
  'topType',
  'hairColor',
  'hatColor',
  'facialHairType',
  'facialHairColor',
  'accessoriesType',
  'clotheType',
  'clotheColor',
];

export const labels = {
  skinColor: 'Skin Colour',
  topType: 'Hair or Headware',
  hairColor: 'Hair Colour',
  hatColor: 'Headware Colour',
  facialHairType: 'Facial Hair Type',
  facialHairColor: 'Facial Hair Colour',
  accessoriesType: 'Glasses',
  clotheType: 'Clothing Type',
  clotheColor: 'Clothing Colour',
};

export const isEditable = {
  skinColor: () => true,
  topType: () => true,
  hairColor: options => [
    'NoHair',
    'Eyepatch',
    'Hat',
    'Hijab',
    'Turban',
    'WinterHat1',
    'WinterHat2',
    'WinterHat3',
    'WinterHat4',
    'LongHairFrida',
    'LongHairShavedSides',
  ].indexOf(options.topType) < 0,
  hatColor: options => [
    'Hijab',
    'Turban',
    'WinterHat1',
    'WinterHat2',
    'WinterHat3',
    'WinterHat4',
  ].indexOf(options.topType) >= 0,
  facialHairType: options => options.topType !== 'Hijab',
  facialHairColor: options => isEditable.facialHairType(options) && options.facialHairType !== 'Blank',
  accessoriesType: () => true,
  clotheType: () => true,
  clotheColor: options => !options.clotheType.startsWith('Blazer'),
};

export const DefaultAvatar = {
  accessoriesType: 'Blank',
  avatarStyle: 'Transparent',
  clotheColor: 'Blue03',
  clotheType: 'BlazerShirt',
  eyeType: 'Default',
  eyebrowType: 'Default',
  facialHairColor: 'BrownDark',
  facialHairType: 'Blank',
  hairColor: 'BrownDark',
  mouthType: 'Twinkle',
  skinColor: 'Light',
  topType: 'LongHairStraight',
};
