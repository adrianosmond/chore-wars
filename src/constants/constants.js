const TIME_UNIT = 1000 * 60 * 60 * 24; // 1 day
const MIN_CHORE_FREQUENCY = 1;
const MAX_CHORE_FREQUENCY = 365;
const MIN_CHORE_POINTS = 1;
const MAX_CHORE_POINTS = 200;

const DATE_FORMAT = 'DD/MM/YYYY';

const MAX_POINT_DIFFERENCE = 500;

const MAX_NAME_LENGTH = 10;
const JOIN_CODE_LENGTH = 8;

const DefaultAvatar = {
  accessoriesType: 'Blank',
  avatarStyle: 'Transparent',
  clotheColor: 'Blue03',
  clotheType: 'BlazerShirt',
  eyeType: 'Default',
  eyebrowType: 'Default',
  facialHairColor: 'BrownDark',
  facialHairType: 'Blank',
  hairColor: 'BrownDark',
  mouthType: 'Twinkle',
  skinColor: 'Light',
  topType: 'LongHairStraight',
};

const ActionTypes = {
  addChore: 'ADD_CHORE',
  resetChoreDoneDate: 'RESET_CHORE_DONE_DATE',
  blockChore: 'BLOCK_CHORE',
  unblockChore: 'UNBLOCK_CHORE',
  removeChore: 'REMOVE_CHORE',
  updateChore: 'UPDATE_CHORE',
  setChores: 'SET_CHORES',
  makeChain: 'MAKE_CHAIN',
  breakChain: 'BREAK_CHAIN',
  addPoints: 'ADD_POINTS',
  claimPrize: 'CLAIM_PRIZE',
  paidDebt: 'PAID_DEBT',
  setPoints: 'SET_POINTS',
  updateUser: 'UPDATE_USER',
  setAuthUser: 'SET_AUTH_USER',
  setGame: 'SET_GAME',
  setPointsLoaded: 'SET_POINTS_LOADED',
  setChoresLoaded: 'SET_CHORES_LOADED',
  saveStatePostUndo: 'SAVE_STATE_POST_UNDO',
};

export {
  TIME_UNIT,
  MIN_CHORE_FREQUENCY,
  MAX_CHORE_FREQUENCY,
  MIN_CHORE_POINTS,
  MAX_CHORE_POINTS,
  MAX_NAME_LENGTH,
  JOIN_CODE_LENGTH,
  DATE_FORMAT,
  MAX_POINT_DIFFERENCE,
  ActionTypes,
  DefaultAvatar,
};
