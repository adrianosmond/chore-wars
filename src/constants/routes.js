export const LOGIN = '/'
export const CHORES = '/chores'
export const NEW_CHORE = '/chores/new'
