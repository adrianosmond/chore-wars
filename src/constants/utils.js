import { TIME_UNIT } from './constants'
const sortByCurrentPoints = (a, b) => b.currentPoints - a.currentPoints

const processChore = (chore) => {
  const timeSinceChore = (new Date().getTime() - chore.lastDone) / TIME_UNIT
  const timeRemaining = timeSinceChore - chore.frequency
  const percentage = Math.min(100 * timeSinceChore / chore.frequency, 100)
  const multiplier = 1 + (1 / chore.frequency)
  const due = chore.lastDone + (chore.frequency * TIME_UNIT)
  return {
    currentPoints: Math.round(percentage / 100 * chore.pointsPerTime * Math.pow(multiplier, timeRemaining)),
    percentage,
    due
  }
}

export { sortByCurrentPoints, processChore }
