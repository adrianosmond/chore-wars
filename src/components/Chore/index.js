import React from 'react';
import { connect } from 'react-redux';
import fecha from 'fecha';

import PopUpMenu from '../PopUpMenu';

import { completeChore, removeChore, breakChain } from '../../actions/choreActions';
import { DATE_FORMAT } from '../../constants/constants';
import * as routes from '../../constants/routes';

import './index.css';

import checkIcon from '../../images/check.svg';

const Chore = ({
  chore, user, game, markChoreComplete, deleteChore, breakChoreChain,
}) => {
  const bonusPoints = chore.currentPoints > chore.pointsPerTime;
  const timeSinceChore = (new Date().getTime() - chore.lastDone);
  /* 
    {canBreak ? <button className="chore-menu__item js-break-chain" onClick={this.breakChain.bind(this)}>Break chain</button> : null}
  */
  const menuOptions = [
    { type: 'link', to: `${routes.EDIT_CHORE}/${chore.slug}`, text: 'Edit' },
    { type: 'button', onClick: () => deleteChore(game, chore.slug), text: 'Delete' },
    { type: 'link', to: `${routes.LOG_PAST_COMPLETION}/${chore.slug}`, text: 'I fogot to log this' },
  ];
  if (chore.enables) {
    menuOptions.push({
      type: 'button',
      onClick: () => breakChoreChain(game, chore.slug),
      text: 'Break Chain',
    });
  }
  return (
    <div className="chore">
      <span className="chore__title">{chore.title}</span>
      <div className="chore__bar">
        <div className={`chore__bar-inner${bonusPoints ? ' chore__bar-inner--bonus' : ''}`} style={{
          width: `${chore.percentage}%`,
        }}>
        <span className={`chore__points${bonusPoints ? ' chore__points--bonus' : ''}`}>{chore.currentPoints}</span>
      </div>
    </div>
    <div className="chore__last-done">
      Done: {fecha.format(new Date(chore.lastDone), DATE_FORMAT)}<br/>
      { chore.frequency > 0 ?
        <span>Due: <span className={`${bonusPoints ? 'chore__last-done--overdue' : ''}`}>{fecha.format(new Date(chore.due), DATE_FORMAT)}</span></span>
        : null }
    </div>
    <button className="chore__complete-button"
      onClick={markChoreComplete.bind(null, chore, user, game)}
      disabled={timeSinceChore < 60000}>
      <img src={checkIcon} alt="Mark as complete" />
    </button>
    <PopUpMenu extraClasses="chore__extra-options"
      options={menuOptions}>
      <div className="chore__extra-icon">
        <span className="chore__extra-icon-text">â€¦</span>
      </div>
    </PopUpMenu>
  </div>
  );
};

const mapStateToProps = state => ({
  user: state.session.authUser.uid,
  game: state.session.game.gameId,
});

const mapDispatchToProps = dispatch => ({
  markChoreComplete: (chore, user, game) => dispatch(completeChore(chore, user, game)),
  deleteChore: (game, slug) => dispatch(removeChore(game, slug)),
  breakChoreChain: (game, slug) => dispatch(breakChain(game, slug)),
});

export { Chore };

export default connect(mapStateToProps, mapDispatchToProps)(Chore);
