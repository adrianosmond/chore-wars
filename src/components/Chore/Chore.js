import React from 'react';
import fecha from 'fecha';
import PropTypes from 'prop-types';

import PopUpMenu from 'components/PopUpMenu';

import { completeChore, removeChore, breakChain } from 'utils/database';
import { DATE_FORMAT } from 'constants/constants';
import * as routes from 'constants/routes';

import checkIcon from 'images/check.svg';
import './Chore.css';

const Chore = ({
  chore, user, game, allChores,
}) => {
  const bonusPoints = chore.currentPoints > chore.pointsPerTime;
  const menuOptions = [
    { type: 'link', to: `${routes.EDIT_CHORE}/${chore.slug}`, text: 'Edit' },
    { type: 'button', onClick: () => removeChore(game, chore.slug, allChores), text: 'Delete' },
    { type: 'link', to: `${routes.LOG_PAST_COMPLETION}/${chore.slug}`, text: 'I forgot to log this' },
  ];
  if (chore.enables) {
    menuOptions.push({
      type: 'button',
      onClick: () => breakChain(game, chore.slug, allChores),
      text: 'Break Chain',
    });
  }
  return (
    <div className="chore">
      <span className="chore__title">{chore.title}</span>
      <div className="chore__bar">
        <div
          className={`chore__bar-inner${bonusPoints ? ' chore__bar-inner--bonus' : ''}`}
          style={{
            width: `${chore.percentage}%`,
          }}
        >
          <span className={`chore__points${bonusPoints ? ' chore__points--bonus' : ''}`}>
            {chore.currentPoints}
          </span>
        </div>
      </div>
      <div className="chore__last-done">
        {'Done: '}
        {fecha.format(new Date(chore.lastDone), DATE_FORMAT)}
        <br />
        { chore.frequency > 0
          ? (
            <span>
              {'Due: '}
              <span className={`${bonusPoints ? 'chore__last-done--overdue' : ''}`}>
                {fecha.format(new Date(chore.due), DATE_FORMAT)}
              </span>
            </span>
          )
          : null }
      </div>
      <button
        type="button"
        className="chore__complete-button"
        onClick={() => completeChore(chore, user, game)}
      >
        <img src={checkIcon} alt="Mark as complete" />
      </button>
      <PopUpMenu
        extraClasses="chore__extra-options"
        options={menuOptions}
      >
        <div className="chore__extra-icon">
          <span className="chore__extra-icon-text">â€¦</span>
        </div>
      </PopUpMenu>
    </div>
  );
};

Chore.propTypes = {
  chore: PropTypes.objectOf(PropTypes.any).isRequired,
  user: PropTypes.string.isRequired,
  game: PropTypes.string.isRequired,
  allChores: PropTypes.objectOf(PropTypes.any).isRequired,
};

export default Chore;
