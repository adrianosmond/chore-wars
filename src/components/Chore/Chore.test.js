import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import { processChore, makeSlug } from 'constants/utils';
import Chore from './Chore';

const choreTime = new Date('2018-02-18').getTime();
const chore = {
  frequency: 1,
  lastDone: choreTime,
  pointsPerTime: 100,
  title: 'Test Chore',
};
const slug = makeSlug(chore.title);
const processedChore = processChore(chore, slug, choreTime);
const choreProps = {
  chore: processedChore,
};

const mockStore = configureMockStore([thunk]);
const store = mockStore({
  chores: {
    [slug]: chore,
  },
  session: {
    authUser: {
      uid: 'test-user',
    },
    game: {
      gameId: 'test-game',
    },
  },
});

describe('Chore', () => {
  it('Renders a basic chore', () => {
    expect(shallow(<Chore store={store} { ...choreProps } />).dive()).toMatchSnapshot();
  });

  it('Renders a chore with frequency 0', () => {
    const props = {
      ...choreProps,
      chore: processChore({
        ...chore,
        frequency: 0,
      }, slug, choreTime),
    };
    expect(shallow(<Chore store={store} { ...props } />).dive()).toMatchSnapshot();
  });

  it('Should handle clicks on the complete button', () => {
    const mountedChore = shallow(<Chore store={store} { ...choreProps } />).dive();

    const completeFn = jest.fn();
    mountedChore.setProps({
      markChoreComplete: completeFn,
    });

    mountedChore.find('.chore__complete-button').simulate('click');
    expect(completeFn).toHaveBeenCalled();
  });
});
