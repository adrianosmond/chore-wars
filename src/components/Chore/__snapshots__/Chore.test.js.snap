// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chore Renders a basic chore 1`] = `
<div
  className="chore"
>
  <span
    className="chore__title"
  >
    Test Chore
  </span>
  <div
    className="chore__bar"
  >
    <div
      className="chore__bar-inner"
      style={
        Object {
          "width": "0%",
        }
      }
    >
      <span
        className="chore__points"
      >
        0
      </span>
    </div>
  </div>
  <div
    className="chore__last-done"
  >
    Done: 
    18/02/2018
    <br />
    <span>
      Due: 
      <span
        className=""
      >
        19/02/2018
      </span>
    </span>
  </div>
  <button
    className="chore__complete-button"
    disabled={false}
    onClick={[Function]}
  >
    <img
      alt="Mark as complete"
      src="check.svg"
    />
  </button>
  <PopUpMenu
    extraClasses="chore__extra-options"
    options={
      Array [
        Object {
          "text": "Edit",
          "to": "/chores/edit/test-chore",
          "type": "link",
        },
        Object {
          "onClick": [Function],
          "text": "Delete",
          "type": "button",
        },
        Object {
          "text": "I fogot to log this",
          "to": "/chores/log/test-chore",
          "type": "link",
        },
      ]
    }
  >
    <div
      className="chore__extra-icon"
    >
      <span
        className="chore__extra-icon-text"
      >
        …
      </span>
    </div>
  </PopUpMenu>
</div>
`;

exports[`Chore Renders a chore with frequency 0 1`] = `
<div
  className="chore"
>
  <span
    className="chore__title"
  >
    Test Chore
  </span>
  <div
    className="chore__bar"
  >
    <div
      className="chore__bar-inner"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <span
        className="chore__points"
      >
        100
      </span>
    </div>
  </div>
  <div
    className="chore__last-done"
  >
    Done: 
    18/02/2018
    <br />
  </div>
  <button
    className="chore__complete-button"
    disabled={false}
    onClick={[Function]}
  >
    <img
      alt="Mark as complete"
      src="check.svg"
    />
  </button>
  <PopUpMenu
    extraClasses="chore__extra-options"
    options={
      Array [
        Object {
          "text": "Edit",
          "to": "/chores/edit/test-chore",
          "type": "link",
        },
        Object {
          "onClick": [Function],
          "text": "Delete",
          "type": "button",
        },
        Object {
          "text": "I fogot to log this",
          "to": "/chores/log/test-chore",
          "type": "link",
        },
      ]
    }
  >
    <div
      className="chore__extra-icon"
    >
      <span
        className="chore__extra-icon-text"
      >
        …
      </span>
    </div>
  </PopUpMenu>
</div>
`;
